File: src/lib/embedding.ts
/**
 * Unified embedding provider for TypeScript (route handlers).
 * Mirrors the Python agents/base/embedding.py abstraction.
 */

const TIMEOUT_MS = 30_000;
const BATCH_LIMIT = 10;
const DEFAULT_DIMENSIONS = 1024;

interface EmbeddingProvider {
  embed(texts: string[]): Promise<(number[] | null)[]>;
}

export interface EmbeddingRuntimeConfig {
  provider?: string;
  apiKey?: string;
  baseUrl?: string;
  model?: string;
  dimensions?: number;
}

function getConfig() {
  return {
    provider: process.env.EMBEDDING_PROVIDER || "openai",
    apiKey: process.env.EMBEDDING_API_KEY || process.env.LLM_API_KEY || "",
    baseUrl:
      process.env.EMBEDDING_BASE_URL ||
      process.env.LLM_BASE_URL ||
      "https://api.openai.com/v1",
    model: process.env.EMBEDDING_MODEL || "text-embedding-3-small",
    dimensions: parseInt(
      process.env.EMBEDDING_DIMENSIONS || String(DEFAULT_DIMENSIONS),
      10
    ),
  };
}

async function openaiEmbed(
  texts: string[],
  config: ReturnType<typeof getConfig>
): Promise<(number[] | null)[]> {
  const results: (number[] | null)[] = [];

  for (let i = 0; i < texts.length; i += BATCH_LIMIT) {
    const batch = texts.slice(i, i + BATCH_LIMIT);
    try {
      const controller = new AbortController();
      const timer = setTimeout(() => controller.abort(), TIMEOUT_MS);

      const resp = await fetch(`${config.baseUrl}/embeddings`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${config.apiKey}`,
        },
        body: JSON.stringify({
          model: config.model,
          input: batch,
          dimensions: config.dimensions,
        }),
        signal: controller.signal,
      });

      clearTimeout(timer);

      if (!resp.ok) {
        const errText = await resp.text().catch(() => "unknown error");
        console.warn(`Embedding API error (${resp.status}): ${errText}`);
        results.push(...batch.map(() => null));
        continue;
      }

      const json = await resp.json();
      for (const item of json.data) {
        results.push(item.embedding as number[]);
      }
    } catch (err) {
      console.warn("Embedding batch failed:", err);
      results.push(...batch.map(() => null));
    }
  }

  return results;
}

function resolveBaseUrl(
  provider: string,
  config: ReturnType<typeof getConfig>
): string {
  if (config.baseUrl) return config.baseUrl;

  switch (provider) {
    case "dashscope":
      return "https://dashscope.aliyuncs.com/compatible-mode/v1";
    case "zhipu":
      return "https://open.bigmodel.cn/api/paas/v4/";
    default:
      return config.baseUrl;
  }
}

function resolveModel(
  provider: string,
  config: ReturnType<typeof getConfig>
): string {
  if (config.model) return config.model;

  switch (provider) {
    case "dashscope":
      return "text-embedding-v4";
    case "zhipu":
      return "embedding-3";
    default:
      return "text-embedding-3-small";
  }
}

let _cached: EmbeddingProvider | null = null;

function buildEmbeddingProvider(
  override?: EmbeddingRuntimeConfig
): EmbeddingProvider {
  const mergedBase = getConfig();
  const config = {
    ...mergedBase,
    ...override,
    provider: (override?.provider || mergedBase.provider || "openai").toLowerCase(),
    dimensions:
      typeof override?.dimensions === "number"
        ? override.dimensions
        : mergedBase.dimensions,
  };
  const provider = config.provider.toLowerCase();

  if (provider === "local") {
    console.warn(
      "Local embedding not supported on TS side; falling back to API-based."
    );
  }

  const resolved = {
    ...config,
    baseUrl: resolveBaseUrl(provider, config),
    model: resolveModel(provider, config),
  };

  return {
    embed: (texts: string[]) => openaiEmbed(texts, resolved),
  };
}

function getEmbeddingProvider(override?: EmbeddingRuntimeConfig): EmbeddingProvider {
  if (override) {
    return buildEmbeddingProvider(override);
  }
  if (_cached) return _cached;
  _cached = buildEmbeddingProvider();
  return _cached;
}

/**
 * Convenience: embed a list of texts using the configured provider.
 * Returns null for any text that failed to embed.
 */
export async function embedTexts(
  texts: string[],
  override?: EmbeddingRuntimeConfig
): Promise<(number[] | null)[]> {
  return getEmbeddingProvider(override).embed(texts);
}

/**
 * Convenience: embed a single text.
 */
export async function embedText(
  text: string,
  override?: EmbeddingRuntimeConfig
): Promise<number[] | null> {
  const results = await embedTexts([text], override);
  return results[0] ?? null;
}

export { getEmbeddingProvider, type EmbeddingProvider };

File: agents/base/embedding.py
"""Unified embedding provider abstraction.

Supports OpenAI, DashScope (Qwen), Zhipu, and local BGE models.
Provider is selected via EMBEDDING_PROVIDER env var.
"""

from __future__ import annotations

import logging
import os
import time
from abc import ABC, abstractmethod
from typing import Optional

logger = logging.getLogger("embedding")

BATCH_LIMIT = 10
TIMEOUT_SECONDS = 30
DEFAULT_DIMENSIONS = 1024


def _get_dimensions() -> int:
    val = os.environ.get("EMBEDDING_DIMENSIONS")
    if val:
        try:
            return int(val)
        except ValueError:
            pass
    return DEFAULT_DIMENSIONS


class EmbeddingProvider(ABC):
    """Base class for all embedding providers."""

    @abstractmethod
    def embed(self, texts: list[str]) -> list[list[float] | None]:
        """Return a list of embedding vectors (or None on per-item failure)."""
        ...

    def embed_single(self, text: str) -> list[float] | None:
        results = self.embed([text])
        return results[0] if results else None


class OpenAIEmbedding(EmbeddingProvider):
    """OpenAI text-embedding-3-small (or compatible API like DashScope)."""

    def __init__(self) -> None:
        from openai import OpenAI

        api_key = os.environ.get("EMBEDDING_API_KEY") or os.environ.get("LLM_API_KEY", "")
        base_url = os.environ.get("EMBEDDING_BASE_URL") or os.environ.get("LLM_BASE_URL", "https://api.openai.com/v1")
        self._model = os.environ.get("EMBEDDING_MODEL", "text-embedding-3-small")
        self._dims = _get_dimensions()
        self._client = OpenAI(api_key=api_key, base_url=base_url, timeout=TIMEOUT_SECONDS)

    def embed(self, texts: list[str]) -> list[list[float] | None]:
        results: list[list[float] | None] = []
        for i in range(0, len(texts), BATCH_LIMIT):
            batch = texts[i : i + BATCH_LIMIT]
            try:
                resp = self._client.embeddings.create(
                    model=self._model,
                    input=batch,
                    dimensions=self._dims,
                )
                for item in resp.data:
                    results.append(item.embedding)
            except Exception as e:
                logger.warning("OpenAI embedding batch failed: %s", e)
                results.extend([None] * len(batch))
        return results


class DashScopeEmbedding(EmbeddingProvider):
    """DashScope (Qwen) embedding via OpenAI-compatible API."""

    def __init__(self) -> None:
        from openai import OpenAI

        api_key = os.environ.get("EMBEDDING_API_KEY") or os.environ.get("LLM_API_KEY", "")
        base_url = os.environ.get("EMBEDDING_BASE_URL", "https://dashscope.aliyuncs.com/compatible-mode/v1")
        self._model = os.environ.get("EMBEDDING_MODEL", "text-embedding-v4")
        self._dims = _get_dimensions()
        self._client = OpenAI(api_key=api_key, base_url=base_url, timeout=TIMEOUT_SECONDS)

    def embed(self, texts: list[str]) -> list[list[float] | None]:
        results: list[list[float] | None] = []
        for i in range(0, len(texts), BATCH_LIMIT):
            batch = texts[i : i + BATCH_LIMIT]
            try:
                resp = self._client.embeddings.create(
                    model=self._model,
                    input=batch,
                    dimensions=self._dims,
                )
                for item in resp.data:
                    results.append(item.embedding)
            except Exception as e:
                logger.warning("DashScope embedding batch failed: %s", e)
                results.extend([None] * len(batch))
        return results


class ZhipuEmbedding(EmbeddingProvider):
    """Zhipu embedding-3 via OpenAI-compatible API."""

    def __init__(self) -> None:
        from openai import OpenAI

        api_key = os.environ.get("EMBEDDING_API_KEY", "")
        self._model = os.environ.get("EMBEDDING_MODEL", "embedding-3")
        self._dims = _get_dimensions()
        self._client = OpenAI(
            api_key=api_key,
            base_url="https://open.bigmodel.cn/api/paas/v4/",
            timeout=TIMEOUT_SECONDS,
        )

    def embed(self, texts: list[str]) -> list[list[float] | None]:
        results: list[list[float] | None] = []
        for i in range(0, len(texts), BATCH_LIMIT):
            batch = texts[i : i + BATCH_LIMIT]
            try:
                resp = self._client.embeddings.create(
                    model=self._model,
                    input=batch,
                    dimensions=self._dims,
                )
                for item in resp.data:
                    results.append(item.embedding)
            except Exception as e:
                logger.warning("Zhipu embedding batch failed: %s", e)
                results.extend([None] * len(batch))
        return results


class LocalBGEEmbedding(EmbeddingProvider):
    """Local BGE-small-zh via sentence-transformers."""

    def __init__(self) -> None:
        try:
            from sentence_transformers import SentenceTransformer
        except ImportError:
            raise ImportError("sentence-transformers is required for local embedding. pip install sentence-transformers")

        model_name = os.environ.get("EMBEDDING_MODEL", "BAAI/bge-small-zh-v1.5")
        logger.info("Loading local embedding model: %s", model_name)
        start = time.time()
        self._model = SentenceTransformer(model_name)
        logger.info("Model loaded in %.1fs", time.time() - start)

    def embed(self, texts: list[str]) -> list[list[float] | None]:
        try:
            vectors = self._model.encode(texts, normalize_embeddings=True)
            return [v.tolist() for v in vectors]
        except Exception as e:
            logger.warning("Local embedding failed: %s", e)
            return [None] * len(texts)


_provider_cache: Optional[EmbeddingProvider] = None


def get_embedding_provider() -> EmbeddingProvider:
    """Factory: return the configured EmbeddingProvider singleton."""
    global _provider_cache
    if _provider_cache is not None:
        return _provider_cache

    provider_name = os.environ.get("EMBEDDING_PROVIDER", "openai").lower()

    providers = {
        "openai": OpenAIEmbedding,
        "dashscope": DashScopeEmbedding,
        "zhipu": ZhipuEmbedding,
        "local": LocalBGEEmbedding,
    }

    cls = providers.get(provider_name)
    if cls is None:
        logger.warning("Unknown EMBEDDING_PROVIDER=%s, falling back to openai", provider_name)
        cls = OpenAIEmbedding

    try:
        _provider_cache = cls()
    except Exception as e:
        logger.error("Failed to initialize embedding provider %s: %s", provider_name, e)
        raise

    return _provider_cache
